core/src/main/java/io/grpc/internal/BinaryLogProvider.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ClientInterceptor;
+import io.grpc.ServerInterceptor;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+public abstract class BinaryLogProvider {
+  private static final Logger logger = Logger.getLogger(BinaryLogProvider.class.getName());
+  private static final BinaryLogProvider NULL_PROVIDER = new NullProvider();
+  private static final BinaryLogProvider PROVIDER = load(BinaryLogProvider.class.getClassLoader());
+
+  /**
+   * Always returns a {@code BinaryLogProvider}, even if the provider always returns null
+   * interceptors.
+   */
+  public static BinaryLogProvider provider() {
+    return PROVIDER;
+  }
+
+  @VisibleForTesting
+  static BinaryLogProvider load(ClassLoader classLoader) {
+    try {
+      return loadHelper(classLoader);
+    } catch (Throwable t) {
+      logger.log(
+          Level.SEVERE, "caught exception loading BinaryLogProvider, will disable binary log", t);
+      return NULL_PROVIDER;
+    }
+  }
+
+  private static BinaryLogProvider loadHelper(ClassLoader classLoader) {
+    Iterator<BinaryLogProvider> iter = getCandidatesViaServiceLoader(classLoader).iterator();
+    List<BinaryLogProvider> list = new ArrayList<BinaryLogProvider>();
+    while (iter.hasNext()) {
+      // The iterator comes from ServiceLoader and may throw when next() is called
+      try {
+        list.add(iter.next());
+      } catch (ServiceConfigurationError e) {
+        logger.log(Level.SEVERE, "caught exception creating an instance of BinaryLogProvider", e);
+      }
+    }
+    if (list.isEmpty()) {
+      return NULL_PROVIDER;
+    } else {
+      return Collections.max(list, new Comparator<BinaryLogProvider>() {
+        @Override
+        public int compare(BinaryLogProvider f1, BinaryLogProvider f2) {
+          return f1.priority() - f2.priority();
+        }
+      });
+    }
+  }
+
+  private static ServiceLoader<BinaryLogProvider> getCandidatesViaServiceLoader(
+      ClassLoader classLoader) {
+    ServiceLoader<BinaryLogProvider> i = ServiceLoader.load(BinaryLogProvider.class, classLoader);
```

Comment:
This is _very_ expensive on Android because it requires loading/validating the apk zip and then it can leave considerable amount of memory left-over for the zip table. If we detect Android we could just use NULL_PROVIDER.